libraries/AP_SmartAudio/AP_SmartAudio.cpp

```diff
@@ -0,0 +1,955 @@
+    // @User: Advanced
+    AP_GROUPINFO_FLAGS("ENABLED",     1, AP_SmartAudio::configured_smartaudio_params, enabled, 0,AP_PARAM_FLAG_ENABLE),
+
+    // @Param: VERSION
+    // @DisplayName: SmartAudio version
+    // @Description: SmartAudio version
+    // @Values: 0: Version 1,1:Version 2, 3: Version 2.1
+    // @User: Advanced
+    AP_GROUPINFO("VERSION",     2, AP_SmartAudio::configured_smartaudio_params, protocol_version, 0),
+
+    // @Param: DEFAULTS
+    // @DisplayName: VTX will be configured with VTX params defined by default.
+    // @Description: VTX will be configured with VTX params defined by default.
+    // @Values: 0: Disabled,1: Enabled
+    // @User: Advanced
+    AP_GROUPINFO("DEFAULTS",     3, AP_SmartAudio::configured_smartaudio_params, setup_defaults, 0),
+
+    AP_GROUPEND
+
+};
+/**
+ * Constructor
+ */
+AP_SmartAudio::AP_SmartAudio()
+{
+    AP_Param::setup_object_defaults(this, var_info);
+    singleton = this;
+
+}
+
+AP_SmartAudio *AP_SmartAudio::singleton;
+AP_VideoTX *ap_video_tx=AP_VideoTX::get_singleton();
+
+// initialization start making a request settings to the vtx
+bool AP_SmartAudio::init()
+{
+        
+        debug("SmartAudio init");
+
+    if(smartaudio_params.enabled.get()==0){
+        
+        debug("SmartAudio protocol it's not active");
+        return false;
+    }
+        
+        debug("SmartAudio protocol is active");
+
+    // init uart
+    _port = AP::serialmanager().find_serial(AP_SerialManager::SerialProtocol_SmartAudio, 0);
+    if(_port!=nullptr){
+
+          if (!hal.scheduler->thread_create(FUNCTOR_BIND_MEMBER(&AP_SmartAudio::loop, void),
+                                          "SmartAudio",
+                                          1024, AP_HAL::Scheduler::PRIORITY_IO, 60)) {
+            return false;
+            }
+
+        // setup port options
+        _port->set_stop_bits(2);
+        _port->set_flow_control(AP_HAL::UARTDriver::FLOW_CONTROL_DISABLE);
+        _port->set_options(AP_HAL::UARTDriver::OPTION_HDPLEX);
+
+        
+        // initialize AP_VideoTx settings
+        request_settings();      
+
+        return true;
+    }
+    return false;
+}
+
+void AP_SmartAudio::loop(){
+
+    // initialise uart (this must be called from within tick b/c the UART begin must be called from the same thread as it is used from)
+    _port->begin(AP_SERIALMANAGER_SMARTAUDIO_BAUD, AP_SERIALMANAGER_SMARTAUDIO_BUFSIZE_RX, AP_SERIALMANAGER_SMARTAUDIO_BUFSIZE_TX);
+
+
+     while (true) {
+
+        // now time to control loop switching
+        uint32_t _now=AP_HAL::millis();
+
+        // when one packet is processed in the iteration will be true.
+        bool packet_processed=false;
+
+        // setup a queue polling delay of 20 ms.
+        hal.scheduler->delay(50);
+
+        // command to process
+        packet _current_command;
+        
+        // Proccess response in the next 200 milis from the request are sent.
+        debug("LOOP: Checking Responses");
```

Comment:
Cut down the debug - difficult to review like this. Probably need to think about the sleep interval and also what happens when we are armed.